project01 peer review PR

jacque-caldwell opened the pull request:
README, stuff in here

jacque-caldwell committed:
Create README.md

README, stuff in here

jacque-caldwell committed:
first test checkin

Does not work/run, just pseudo code, in the wrong area of the file.

jacque-caldwell committed:
updated README to have requirements for project 01

started README file with proper formatting

jacque-caldwell committed:
Add files via upload

jacque-caldwell committed:
Delete README.md

jacque-caldwell committed:
Create project01.py

My solution to project01 homework

jacque-caldwell committed:
Update project01.Rmd

jacque-caldwell committed:
Add files via upload

More tweaks, changes to documentation

jacque-caldwell committed:
text edits

jacque-caldwell committed:
fixed formatting problem with Rmd

jacque-caldwell committed:
Update README.Rmd

jacque-caldwell committed:
Add files via upload

jacque-caldwell committed:
Add files via upload

Candy-Train commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,331 @@
+included by libraries already, or were similar to read file functions that would 
+populate entire data frames in a single read function.  I have done parsing and 
+reading of files in C, and know what I want to do, just not always the way to 
+do it in python.   I found myself looking up a lot of stuff, but I expect this to 
+get better over time.
+2. No/little knowledge of programming,debugging in IDEs RStudio or Visual Studio 
+Code.  Prior python coding was done exclusively on google Colab.   Still working 
+on this, but I have decided that I will be working with Visual Studio Code as an 
+IDE, and somehow moving the code over to the RStudio program on NU cluster system.
+3. No prior knowledge of computer cluster use, no prior use of submitting scripts 
```
I had similar challenges. My Python is a bit rusty, so I knew that things I wanted to do were possible, but still had to look up specific function names to implement them. Likewise, I've also done my previous programing in Visual Studio, not RStudio. While I had the same issue with the debugger in RStudio, I think in the long run, it will be more beneficial for us in this class to learn how to use it rather than jumping between IDEs.

Candy-Train commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,331 @@
+input: list of clinical states from the most recent parse of a line from the 
+CLNDN key-value pair in our VCF file.
+
+internal variables used:  my_len -- size of list sent to the function.
+                          new_list -- list being created to return from this 
+                          function.
+                          item -- is also used as part of a for loop to cyle 
+                          through the mylist "list"
```
This is helpful! When I write code, I've only included the purpose of the function and its inputs/outputs. Including a reference of the variables it uses and what they do is a good idea. Building off of that, it might be a good idea to include what type of variable it is, just for additional clarity. Just looking at this, I'm not entirely sure what `item` is supposed to be.

jacque-caldwell replied:
Agreed. @michael-bambha

Candy-Train commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,331 @@
+issues:  No specific concerns here, did error check for empty lists, or lists 
+with empty strings included in their lists.
+'''
+def clean_my_list(mylist):
```
It seems like the two ways people approached this was to either never include `not_specified` and `not_provided` in the first place, or to do like you did and cut it out at the end. I'm not sure if one way is really any better than another.

jacque-caldwell replied:
As long as it is only done during the parsing of a line with AF_EXAC in it, or with the final list, I can't think of why it would be faster or more efficient one way or the other.

Candy-Train commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,331 @@
+to see if the allele frequency is rare.   If it is, the function calls 
+"find_clinical diseases" and returns the list.
+
+input:  str1 - line of VCF file (one that is not the header or a comment line)
```
I would be careful using vague variable names. While you specify what the variable is here, in larger programs it can be easy to lose track of what something like `str1` is. It also makes it easier to accidentally use the same variable name twice, and cause unintended interference.

jacque-caldwell replied:
Agreed.  I will do better.

Candy-Train commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,331 @@
+key-value string, however, I noted by looking at my output that there were a 
+couple of strings that were longer, and seemed to be using a ',' separator as
+well.  
+2. string.split behviour only uses a single character; I needed to be able to 
+split using two different characters, so I ended up importing the package re in 
+order to help me do this.
```
Would you have been able to use `split` twice to achieve the same result? `split` first using "|', then use `split` on the result to split it by ","?

Candy-Train commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,331 @@
+apparently the way to do this correctly is to surround the string with braces.  
+This took a while to figure out.
+'''
+def read_file(file_name_str):
+  disease_list=list() # initiallize 
+  with open(file_name_str,"r") as fp:
+    for line in fp:
+      if (line[0] == ""):
+        return (disease_list)
+      if (line[0] == "#"):
+        continue # ignore header lines in file
+      else:
+        temp_list = parse_line(line) # parse the line 
+        if temp_list:                 # if clinical disease states are returned, then...
+          if disease_list==0:
+            disease_list = temp_list # first line without # character at start; begin list
+          else:
+            for i_str in temp_list:
+              if i_str in disease_list: 
+                continue # remove the DUPLICATES
+              else:
+                disease_list.extend([i_str])
+               # disease_list.extend(temp_list) # add the new diseases found
+              continue
+        else:
+          continue
+  return(disease_list) #return created list
```
I'm confused here. It looks like you're only returning a list of diseases at the end, but not tracking the number of occurrences of each disease?  The assignment is "Use a dictionary to count the results given by parse_line to keep a running tally (or count) of the number of times a specific disease is observed".

jacque-caldwell replied:
Unfortunately this is correct.  The project01.Rmd file that I have been working with and thought I had committed has different code altogether in it.  Strings need to be in a dictionary and need to have frequency of use in file added to them.  I happen to know this is the wrong file, as the final version had two references in the appendix section, not two.  sigh.   The file in 'Project01 peer review PR'/projects/project01/project01.Rmd apparently has the code we were submitting that has code with the tally in it.

Candy-Train commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,331 @@
+This took a while to figure out.
+'''
+def read_file(file_name_str):
+  disease_list=list() # initiallize 
```
If all you need here is an empty `list` to start adding to, you can do that with `disease_list = []`

jacque-caldwell replied:
Is disease_list = [] different from disease_list=list()
?????  (see line above?)

Candy-Train reviewed (changes requested):
You seem like you're on the right track. You have a list of the diseases that meet the allele frequency, but I don't see where you're tracking how many times each disease that meets these requirement is observed in the data set. It looks like you return a list at the end, not a dictionary.

zoechow24 commented on project01.py:
```diff
@@ -0,0 +1,205 @@
+  else: return()
+
+'''
+function parse_line() ; takes a string (str1) as input (a full line from the VCF file) ; note that this will not be the 
```
Your explanations helped me follow your code, especially since you used functions that I have never used before. But I think it would be better if you left comments within your code so that it is easier to follow. Some of your explanations can be used as docstrings instead of markdown.

jacque-caldwell replied:
what are docstrings?   do you mean comments?

zoechow24 replied:
docstrings are the documentation enclosed with """ """ or ''' ''' right under the definition line. You have it in the .Rmd file in Project01.py files.

zoechow24 commented on project01.py:
```diff
@@ -0,0 +1,205 @@
+  disease_list=list() # initiallize 
+  with open(file_name_str,"r") as fp:
+    for line in fp:
+      if (line[0] == ""):
+        return (disease_list)
```
I'm confused with this part of your code. Is it supposed to return an empty list if the line is empty?

jacque-caldwell replied:
Nope, this was left in accidentally.  I have removed it, thanks!

zoechow24 commented on project01.py:
```diff
@@ -0,0 +1,205 @@
+challenges:  apparently casting a string to a list creates a list of characters (known and documented behavior, but that seemed a little odd to me; 
+apparently the way to do this correctly is to surround the string with braces. who knew.  This took a while to figure out.
+'''
+def read_file(file_name_str):
```
Your read_file should have returned a dictionary with the counts for diseases instead of a list. I believe your function is creating a list of unique diseases instead?

jacque-caldwell replied:
fixed.  hopefully new file has been resubmitted.  It's too bad, this was not the correct project01.Rmd file that was supposed to be submitted for review, which did have the right answer.

zoechow24 replied:
Sorry! I looked at the wrong files, but your Rmd file looks great. It's concise and easy to follow. I don't really have any comments regarding the code itself. There was a lot I learned from it. Thank you!

zoechow24 reviewed (approved):
The code from Michael is easier to follow due to the short but clear docstrings. However, there are functions you used that I have not learned before so it was definitely very educational, but a bit confusing at times. I think comments can help so that people at different levels can understand what is being done .

jacque-caldwell committed:
Update project01.Rmd

Updated the strange project01.Rmd that had my code and not MB's code in it??   Now returns dictionary with counts instead of just a list of strings.

jacque-caldwell reviewed (commented):
Updated my read_file to return dictionary with counts in it.
This file wasn't my intended Rmd file to check in -- should have been the one with MB's code in it.  Unclear why there are two? project01.Rmd files.
Odd.

e12chaos commented on Project01.py:
```diff
@@ -0,0 +1,111 @@
+"""
+# pylint: disable=invalid-name
+
+from typing import List
+from collections import Counter
+from pprint import pprint
```
I liked that you loaded the libraries at the beginning

e12chaos commented on Project01.py:
```diff
@@ -0,0 +1,111 @@
+
+    attrs = {}
+
+    try:
+        info = vcf_string.split("\t")[7]
+    except (
+        IndexError
+    ):  # in case the line is incorrectly formatted, we return an empty list
+        return []
```
A good way to debug and catch errors using the try/except

e12chaos commented on Project01.py:
```diff
@@ -0,0 +1,111 @@
+    """
+    strings_to_filter = ["not_specified", "not_provided"]
+    clndn_str_filtered = clndn_str.split("|")
+    return [
+        disease for disease in clndn_str_filtered if disease not in strings_to_filter
+    ]
```
Nice job on list comprehension, it's something I'm still working on, but this looks like a good example

e12chaos commented on README.Rmd:
```diff
@@ -0,0 +1,202 @@
+    toc_depth: 2
+    html_preview: false
+---
+
```
The readme is supposed to be a .md, not a .rmd, it looks like you have both and can delete this extra file.

e12chaos commented on project01.py:
```diff
@@ -0,0 +1,205 @@
+            disease_list = temp_list # first line without # character at start; begin list
+          else:
+            for i_str in temp_list:
+              if i_str in disease_list: 
```
I think that each time a new disease is found, Python has to scan through the entire disease_list to check if that disease already exists. As the list grows to hundreds or thousands of entries, this becomes very slow. I think using counter like Michael did could avoid this potential problem

e12chaos commented on Project01.py:
```diff
@@ -0,0 +1,111 @@
+        List[str]: List of diseases found
+    """
+
+    attrs = {}
```
I think that this dictionary-based parsing approach (attrs = {}) is more robust than string searching

brgrhrng commented on projects/project01/README.Rmd:
I think the intent of the README file is just to contain the descriptive text and pseudocode, not your actual project code.

Also, while the assignment doesn't specify a format for this file, I would personally recommend saving it as a plain markdown file (`.md`). If you do, github will automatically show a preview of its contents whenever you navigate to the folder (it's actually currently doing so with the blank `README.md` file.)

brgrhrng commented on projects/project01/README.Rmd:
````diff
@@ -0,0 +1,303 @@
+read_file("clinvar_20190923_short.vcf)
+}
+
+# Python Code block here
+```{python}
````
As mentioned above, I don't think you need to duplicate your python code in this file.

brgrhrng commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,327 @@
+    return 
+
+def read_file(file_name_str):
+  file_pointer = open(file_name_str)
+  
+  while file_pointer not at EOF:
+      next_line = read_next_line(file_pointer) 
+      
+  close(file_pointer)
```
There is nothing wrong with your implementation here--it's efficient, loading the file line-by-line instead of dumping the full thing into memory, and your code (particularly in the `read_next_line()` helper!) is very well-written and easy to follow.

There is an alternate, more "python-y" way to lazily iterate through the lines in a file, without explicitly managing your pointer, checking for the EOF character, etc:
```
with open(filename, 'r') as f:
    for line in f:
       pass # do stuff with line (you often need to .strip() it first)
# python automatically closes the file when it escapes out of the above context
```

brgrhrng commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,327 @@
+to be run overnight, etc.   I have UNIX experience, which does help some, but
+logging in with the OnDemand system confuses me a bit, to know when I'm truely
+logged out.  I worry that I have mutliple sessions that are still running 
+somewhere in the background on one of the servers.   Additionally I do not 
```
I also worry about stuff like this! Any sessions you open through OOD's courses list should be accessible if you click on "my interactive sessions" on the main header. This allows you to re-enter or kill them if they're still running.

brgrhrng commented on projects/project01/project01.Rmd:
```diff
@@ -0,0 +1,327 @@
+                disease_list[i_string]+=1
+        else:
+          continue
+  return(disease_list) #return created list
+        else:
+          continue
```
I'm not sure I understand what these lines are doing. Is `return....else` a valid construction in python?

brgrhrng reviewed (commented):
Overall this project is good, and I don't see any mandatory changes in the code itself. I have noted one point where you might have a syntax error, but it also might just be a python-y construct that I just haven't encountered. Your README hits all the necessary points, and it was interesting to see how similar our groups' challenges were.

You do have several extra code files in your top level directory. I did not add line-item suggestions in these files, since I assume they are accidental uploads. I would highly recommend adding a commit that removes these files (the project leader can do so by navigating to the individual files in github, selecting the three dots, and then selecting 'delete' and then 'commit'). 

Lastly, while I don't think the assignment specifically requires this, I would recommend removing the python code in your README.rmd file and converting it to a plain .md file. Likewise you probably don't need to duplicate your project reflection in project01.Rmd. 

I've graded your project as 'complete' in Canvas.
